### PR TITLE
Jon Peterson

### PR DESCRIPTION
**Pull request by joequincy**

### CODE REVIEW

File: test/house_test.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  def test_house_area_computes_correctly
+    assert_equal 0, @house.area
+    @house.add_room(@room_1)
+    assert_equal 130, @house.area
+    @house.add_room(@room_2)
+    assert_equal 295, @house.area
+    @house.add_room(@room_3)
+    assert_equal 670, @house.area
+    @house.add_room(@room_4)
+    assert_equal 1900, @house.area
```
Comment (BrianZanti) [inline test/house_test.rb]: I think it's a bit over kill to test each one of these. I think testing for zero is a good edge case, and then maybe two others.

File: test/house_test.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  def test_house_calculates_price_per_square_foot
+    assert_nil @house.price_per_square_foot
```
Comment (BrianZanti) [inline test/house_test.rb]: Interesting choice here. Do you want it to return nil? I can see how that would make sense if there are no rooms. In any case, another good example of edge case testing.

File: test/house_test.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  end
+  def test_house_returns_hash_of_rooms_by_category
+    @house.add_room(@room_1)
+    @house.add_room(@room_2)
+    @house.add_room(@room_3)
+    @house.add_room(@room_4)
+    returned_hash = @house.rooms_by_category
+    assert_equal [@room_1,@room_2], returned_hash[:bedroom]
+    assert_equal [@room_3], returned_hash[:living_room]
+    assert_equal [@room_4], returned_hash[:basement]
```
Comment (BrianZanti) [inline test/house_test.rb]: Be as explicit as possible with your expected values. Instead of asserting all these parts of the expected hash, just assert exactly what you want it to be:

```ruby
expected = {
  bedroom: [@room_1, @room_2],l
  living_room: [@room_3],
  basement: [@room_4]
}
```

File: test/house_test.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
+require 'minitest/autorun'
+require 'minitest/pride'
+require 'pry'
+require './lib/room'
+require './lib/house'
+
+class HouseTest < Minitest::Test
```
Comment (BrianZanti) [inline test/house_test.rb]: Testing looks really good 👍

File: lib/room.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+class Room
+  attr_reader :category
```
Comment (BrianZanti) [inline lib/room.rb]: This is subtle, but it's good that you didn't make length and width an attr reader. It wasn't asked for, so need to expose those attributes.

File: lib/house.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  def add_room(room)
+    @rooms << room
+  end
+
+  def rooms_from_category(category)
+    @rooms.select {|room| room.category == category}
+  end
+
+  def area
+    @rooms.inject(0) {|area,room| area += room.area}
```
Comment (BrianZanti) [inline lib/house.rb]: Great job!

File: lib/house.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def price_per_square_foot
+    area == 0 ? nil : (@price.sub("$","").to_f / area).round(2)
```
Comment (BrianZanti) [inline lib/house.rb]: This line is a bit unreadable. Stay away from ternaries as they are less readable than if statements. I would definitely put converting the price on its own line.

File: lib/house.rb [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def rooms_sorted_by_area
+    @rooms.sort {|a,b| b.area <=> a.area}
```
Comment (BrianZanti) [inline lib/house.rb]: Whenever you can, use `sort_by` over sort for readability as the `<=>` operator is less easily understood by other programmers.

```ruby
@rooms.sort_by {|room| room.area}
```

File: lib/house.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  def rooms_by_category
+    output = {}
+    categories = []
+    @rooms.each do |room|
+      if !categories.include?(room.category)
+        categories << room.category
+        output[room.category] = rooms_from_category(room.category)
```
Comment (BrianZanti) [inline lib/house.rb]: This is a creative solution. An alternative approach would be to iterate over the array of rooms and shovel each room into the array associated with its category in the output. Your approach is a bit less efficient since `rooms_from_category` also has to iterate over the array of rooms so you have to do more iterations. Its not quite N squared, but its close. If that doesn't make sense we can talk more about algorithmic complexity in person.

**Comment by BrianZanti** [comment]
Great job!
